[BUGFIX] COLOR-70 - Fixed AOE Spread

SirKingII opened the pull request:
Old version took the condition instance and applied it to multiple enemies, the new instance creates a new condition that ends at the same time and spreads it.

SirKingII committed:
Fixed AOE Spread

Old version took the condition instance and applied it to multiple enemies, the new instance creates a new condition that ends at the same time and spreads it

frenchtr commented on Assets/Runtime/Conditions/DOTCondition.cs:
```diff
@@ -10,7 +10,7 @@ public class DOTCondition : Condition
         /// Damage dealt per tick
         /// </summary>
-        private float Damage;
+        public float Damage;
```
Any reason these are public? I don't see you using this anywhere outside of DOTCondition. Especially since you can pass damage into the constructor already. Making this public breaks this class' encapsulation. The question becomes:

"If I'm another developer looking at your code, how am I supposed to use your class? Do I modify damage directly or do I use the constructor?"

frenchtr replied:
Okay - I see what you're doing here. You want to give public access to Damage and SlowPercent so you can use those values when rebuilding the conditions.

Right now you've exposed your variables as *fields*. That means that any outside code can both *get* and *set* the values, which means that anything can change it and we might not want that behavior. Instead, we probably want to expose a pair of *properties* and expose only the *getter* but not the *setter*. Here's what that looks like:

```
private float slowPercent;

public float SlowPercent
{
        get
        {
              return this.slowPercent;
        }
}
```

This allows us very tight control over the SlowPercent; outside code can read it as needed, but only this code can write the value.

It is a bit verbose though, which is why we have a few other options:

```
private float slowPercent;

public float SlowPercent
{
        get => this.slowPercent;
}
```

...or even more concisely:

```
private float slowPercent;

public float SlowPercent => this.slowPercent;
```

Use at your discretion.

SirKingII replied:
Fixed! I added a new constructor that takes in duration, and I privated those variables, then made custom getters based on this advice!

frenchtr commented on Assets/Runtime/Conditions/SlowCondition.cs:
```diff
@@ -5,7 +5,7 @@ namespace ScringloGames.ColorClash.Runtime.Conditions
     public class SlowCondition : Condition
     {
-        private float slowPercent;
+        public float slowPercent;
```
Same here.

frenchtr replied:
See above.

frenchtr commented on Assets/Runtime/Conditions/AOECondition.cs:
```diff
@@ -45,11 +45,34 @@ public override void OnApplied(ConditionBank bank)
+                            float duration = thisCond.Duration;
+                            Condition newCondition;
+                            // Case uses pattern casting!
+                            switch (thisCond)
```
Very good job for figuring out how to use the switch statement with pattern matching.

Just a small DRY (Don't Repeat Yourself) optimization: you are calling `SetTimeAndApply()` in each of your switch cases. The only thing that's changing in your switch cases is the condition variable. That means you can instead move `SetTimeAndApply()` outside of the switch statement. So it would look more like this:

```
foreach (var thisCond in bank.Conditions)
{
    targetBank.Apply(thisCond);
    float duration = thisCond.Duration;
    Condition newCondition;
    // Case uses pattern casting!
    switch (thisCond)
    {
        case SlowCondition condition:
        {
            newCondition = new SlowCondition(condition.slowPercent);
            break;
        }
        case DOTCondition condition:
        {
            newCondition = new DOTCondition(condition.Damage);
            break;
        }
    }

    SetTimeAndApply(newCondition, condition, targetBank);
}

```

The C# developers realized this is a common thing and gave us a handy little shorthand way of dealing in this situation:

```
foreach (var thisCond in bank.Conditions)
{
    targetBank.Apply(thisCond);
    float duration = thisCond.Duration;
    
    Condition newCondition = (thisCondition) switch {
        SlowCondition condition => new SlowCondition(condition.slowPercent),
        DOTCondition condition => new DOTCondition(condition.Damage),
    };

    SetTimeAndApply(newCondition, condition, targetBank);
}
```

frenchtr replied:
There's also a good chance that this switch pattern thing will become a problem if we introduce lots of conditions, but we'll cross that bridge when we come to it.

SirKingII replied:
I had a reason for this, based on an error that basically said what you said, where if a type wasn't caught, it would try to pass a non-initialized condition down. Rider provided the elegant solution of setting it to null before, and now I use:
```c#
if (newCondition != null) {targetBank.Apply(newCondition)}
``` 
So now, if it changes, it gets caught! And if it doesn't change, it doesn't apply. Now, a bunch of null conditions don't start showing up on enemies.

SirKingII committed:
Changed permissions

Created new, relevant constructors and getters, then turned the variables private.

SirKingII committed:
Reworked getters

Used proper nomenclature and commented getters.

SirKingII committed:
Removed redundant code

First, new condition is null.
Then, switch statement tries to make new condition not null.
Finally, if not null, it applies.

SirKingII commented:
I believe I have addressed all pressed concerns.
